[css-overflow] Implement scrollbar-gutter on viewports

nmoucht opened the pull request:
#### c94f5d66b46909bb9cc72345adf97a6739f36f90
<pre>
[css-overflow] Implement scrollbar-gutter on viewports
<a href="https://bugs.webkit.org/show_bug.cgi?id=257606">https://bugs.webkit.org/show_bug.cgi?id=257606</a>
<a href="https://rdar.apple.com/110468611">rdar://110468611</a>

Reviewed by Simon Fraser.

Updates scrollbar gutter implementation to work for viewports. When setting scrollbar gutter
on the viewport, shrink the size of the visible content rect, and when both-edges is present,
shift the clip layer over like we do when placing the scrollbar on the left.

* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::documentScrollPositionRelativeToViewOrigin const):
(WebCore::ScrollView::verticalScrollbarWidth const):
(WebCore::ScrollView::horizontalScrollbarHeight const):
(WebCore::ScrollView::locationOfContents const):
* Source/WebCore/platform/ScrollView.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::includeVerticalScrollbarSize const):
(WebCore::RenderBox::verticalScrollbarWidth const):

Canonical link: <a href="https://commits.webkit.org/282661@main">https://commits.webkit.org/282661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b53a882270e22083db47bb6e373bd4a37dd71e5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63816 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43173 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16413 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67837 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14424 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65936 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50863 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14704 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51423 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9971 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66885 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39986 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55246 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32040 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36672 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12628 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13297 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58642 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12955 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69533 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7763 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12497 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58745 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7796 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55343 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58887 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6457 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9651 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38993 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40072 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41255 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39815 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/rendering/RenderListBox.h:
```diff
@@ -111,8 +111,8 @@ class RenderListBox final : public RenderBlockFlow, public ScrollableArea {
     virtual bool shouldPanScroll() const { return true; }
     void panScroll(const IntPoint&) override;
 
-    int verticalScrollbarWidth() const override;
```
We should create a No/Yes enum for this parameter, to make the calling side clearer

`enum class ShouldIncludeGutterForDocumentRenderer : bool { No, Yes }`

nmoucht replied:
Removed this.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6188,6 +6193,43 @@ void LocalFrameView::scrollbarWidthChanged(ScrollbarWidth width)
+    return verticalScrollbar()->width();
+}
+
+
```
Extra blank line.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6188,6 +6193,43 @@ void LocalFrameView::scrollbarWidthChanged(ScrollbarWidth width)
+    return scrollbarGutter;
+}
+
+int LocalFrameView::horizontalScrollbarInset() const
```
Easily mis-interpreted as "inset related to the horizontal scrollbar". Maybe `insetForLeftScrollbar()`

nmoucht replied:
Fixed.

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-gutter-root-both-edges-expected.html:
```diff
@@ -0,0 +1,40 @@
+
+<style>
+    html {
+
```
Blank line.

nmoucht replied:
Fixed.

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-gutter-root-both-edges-expected.html:
```diff
@@ -0,0 +1,40 @@
+<!doctype html>
```
We also need the tests with background image and abspos.

nmoucht replied:
This test utilizes abspos (as do other wpts). I don't think there's a good way to do background image as I don't think there's a good way to expose the overhang like using scrollbar-gutter does.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6188,6 +6193,43 @@ void LocalFrameView::scrollbarWidthChanged(ScrollbarWidth width)
+}
+
+
+IntSize LocalFrameView::computeScrollbarGutter() const
```
It's a bit unclear from the function name what this returns. Better to call it `spaceForScrollbarGutters()`.

We really need a term for "scrollbar or scrollbar gutter". Maybe it's "scrollbar space". So this would just be `totalScrollbarSpace()` or something.

nmoucht replied:
Fixed.

nmoucht committed:
[css-overflow] Implement scrollbar-gutter on viewports
https://bugs.webkit.org/show_bug.cgi?id=257606
rdar://110468611

Reviewed by Simon Fraser.

Updates scrollbar gutter implementation to work for viewports. When setting scrollbar gutter
on the viewport, shrink the size of the visible content rect, and when both-edges is present,
shift the clip layer over like we do when placing the scrollbar on the left.

* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::documentScrollPositionRelativeToViewOrigin const):
(WebCore::ScrollView::verticalScrollbarWidth const):
(WebCore::ScrollView::horizontalScrollbarHeight const):
(WebCore::ScrollView::locationOfContents const):
* Source/WebCore/platform/ScrollView.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::includeVerticalScrollbarSize const):
(WebCore::RenderBox::verticalScrollbarWidth const):

Canonical link: https://commits.webkit.org/282661@main
